_posts/2018-11-13-deploy-rails.md

```diff
@@ -0,0 +1,146 @@
+- capistrano/rvm - integration with RVM managed ruby environments
+- capistrano/bundler - integration with bundler
+- capistrano/puma - tasks to manage the puma application server
+
+With cap [added to a rails application](https://www.digitalocean.com/community/tutorials/deploying-a-rails-app-on-ubuntu-14-04-with-capistrano-nginx-and-puma#step-6-%E2%80%94-adding-deployment-configurations-in-the-rails-app), a `cap [environment] deploy` is all we need to run to deploy a new version of our site.
+Capistrano also provides the ability to rollback a deploy if the most recent one breaks.
+That by itself is reason enough to *capify* all the things.
+
+Since cap connects to the server via SSH, we can use it to run any command as we normally would on the server.
+By writing `tasks`, scripts that list the commands to run, both locally and on remote servers, we can automate the process of setting up our server.
+One only needs set up the server once, while extracting the logic out into cap tasks.
+In this way, we can have a tasks to install postgres, another for nginx, one to create a database role, one to update the system, and so on.
+Once you have the scripts together, provisioning a new server stops being a pain. But how do you get those tasks into a different project?
+[Rails application templates](https://guides.rubyonrails.org/rails_application_templates.html) is how.
+
+Rails application templates provide a way to add 'code' to a new or existing rails application.
+The dsl provides actions to add gems and sources to the `Gemfile`, add initializers, rake tasks and arbitrary files.
+We can set rails configuration variables or even run rails and git commands as the template is applied.
+We can use [Thor actions](https://www.rubydoc.info/github/wycats/thor/Thor/Actions#source_paths-instance_method) to fetch and manipulate data from disparate sources for use in our cap tasks.
+To generate a new application and apply the template:
+~~~shell
+rails new [appname] -m /path/to/template.rb # path can also be a url
+~~~
+Or to apply to an existing app:
+~~~shell
+bin/rails app:template LOCATION=/path/to/template.rb # location variable can also be a url
+~~~
+
+#### Example you ask?
+
+[Rails seed](https://github.com/kgathi2/rails_seed) is a Rails Application Template with the lofty goal of getting you up and running a production-ready environment in 15 minutes.
+And it works... mostly. If nothing else, it should give you a more [concrete](https://sw.wikipedia.org/wiki/Zege) image of the gibberish written here.
+The template contains tasks that take care of installing the typical ruby and rails dependencies, setting up postgres, adding firewall rules and most other setup tasks.
+
+The template for now works for Rails 5.0 and is tested on Ubuntu xenial and bionic. It should also work with rails >=5.1 once you [setup encrypted credentials](https://www.starkandwayne.com/blog/rails-5-1-applications-can-be-a-lot-more-secretive-on-cloud-foundry-and-heroku/).
+
```

Comment:
Include "how to" of deploying rails seed on local without using a VPS in the cloud. Talking about vagrant and docker. So basically a step by step for how to use the template to deploy to a docker container or vagrant.